keps/sig-api-machinery/3920-finish-feature-flags/README.md

```diff
@@ -0,0 +1,1015 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubernetes uses feature gates for a wide variety of features. Unfortunately,
+there are many problems (see motivation section) making features unsafe or
+unfit. We describe a desireable state for Kubernetes (see goals) and propose an
+API and supporting libraries and techniques for achieving the goals (see
+design).
+
+## Motivation
+
+A list of problems:
+* It's mostly unsafe to turn a feature on, use it, and turn it back off. This is
+  because state stored while the feature was on is not cleaned up. So instead of
+  going back into a pristine state, the cluster goes into a third state,
+  off-with-dangling-references. This third state is tested less well than either
+  the on or off states.
+* Because of this, it's risky to try alpha features with important clusters.
+* It's not very useful or convenient to make disposable clusters to try alpha
+  features.
+* Because of that, it's extremely hard to get robust feedback about features
+  while they are in alpha and easy to change.
+* There is no place in the code for devs to put turn on / turn off code, even if
+  they wanted to.
+* Likewise there is no place to put upgrade or downgrade code. Many efforts have
+  used built-in resources instead of CRDs simply because there is no place to
+  put CRD install code.
+* It is difficult to know if a feature has been turned on for every relevant
+  binary. Yes, there are metrics, but you have to know which binaries are
+  relevant for what features and your particular cluster's metrics collection
+  system. This problem is especially insidious when there are multiple
+  apiservers in the cluster and they are not configured the same way.
+* It is especially difficult to know if use of a feature at some point makes
+  cluster upgrade unsafe now; for example, an alpha object being written while
+  the feature was briefly on which is incompatable with the default-on beta.
+* It is not easy to test the feature on -> use -> feature off lifecycle.
+* It is not possible to know if a workload with a feature dependency will
+  successfully run on a given cluster.
+
+### Goals
+
+Primary goals:
+* Minimize codebase churn, existing feature authors do little-to-no work
+* Expose the in-use feature states in use in the cluster, and make it easy to
+  read them
+* Make it easy for admins to know when all binaries in the cluster agree about
+  feature states.
```

Comment:
It's a little hard to tell if this is just *surfacing* information, or proposing modifying how we default feature enablement based on that information.

I'm very interested in improving how we deal with scenarios like this:

* a control plane with all nodes on 1.x+ enables features considered stable enough to default on in 1.x (up-to-date clusters get new features asap)
* a control plane with nodes older than 1.x avoids enabling features that require 1.x nodes (slow-moving clusters prioritize safety at the expense of feature velocity)
* a control plane with enabled features that require 1.x nodes handles an attempt to add a node older than 1.x by _____ (rejecting the node? disabling the feature? something else?)
* a control plane with nodes older than 1.x indicates it is not safe to upgrade the control plane beyond supported node/control-plane skew

Reply:
So far this KEP describes surfacing information and the enablement/disablement flow. I need to add the version change flow also.

I think you are asking for some code to run to automatically change a feature state in a direction desired by devs if no preference has been expressed and it is safe to do so, looking at the rest of the cluster. I have not said anything about that in the KEP yet. I think that is possible and desirable but I need to write the version change flow down first.

I think the main requirement from this is that it should be possible to tell if a value is the default, was set on the command line, or was set dynamically (if permitted).

Also if there are multiple such default-changes-due-to-newly-found-safe-conditions, they should not all happen at the same time. But the issue is, what if it breaks the cluster? that will take some designing to get right.

Reply:
Adding a follow-up goal.